libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/ElectronicHorizonObserverImpl.kt

```diff
@@ -0,0 +1,145 @@
+package com.mapbox.navigation.core.trip.session
+
+import com.mapbox.geojson.LineString
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.trip.model.EHorizon
+import com.mapbox.navigation.base.trip.model.EHorizonPosition
+import com.mapbox.navigation.base.trip.model.EHorizonResultType
+import com.mapbox.navigation.base.trip.model.Edge
+import com.mapbox.navigation.base.trip.model.FRC.MOTORWAY
+import com.mapbox.navigation.base.trip.model.FRC.PRIMARY
+import com.mapbox.navigation.base.trip.model.FRC.RESIDENTIAL
+import com.mapbox.navigation.base.trip.model.FRC.SECONDARY
+import com.mapbox.navigation.base.trip.model.FRC.SERVICE_OTHER
+import com.mapbox.navigation.base.trip.model.FRC.TERTIARY
+import com.mapbox.navigation.base.trip.model.FRC.TRUNK
+import com.mapbox.navigation.base.trip.model.FRC.UNCLASSIFIED
+import com.mapbox.navigation.base.trip.model.NameInfo
+import com.mapbox.navigation.utils.internal.JobControl
+import com.mapbox.navigation.utils.internal.ifNonNull
+import com.mapbox.navigator.ElectronicHorizon
+import com.mapbox.navigator.ElectronicHorizonEdge
+import com.mapbox.navigator.ElectronicHorizonObserver
+import com.mapbox.navigator.ElectronicHorizonResultType
+import com.mapbox.navigator.FRC
+import com.mapbox.navigator.GraphPosition
+import com.mapbox.navigator.RoadNameInfo
+import kotlinx.coroutines.launch
+import java.util.concurrent.CopyOnWriteArraySet
+
+internal class ElectronicHorizonObserverImpl(
+    private val jobController: JobControl
+) : ElectronicHorizonObserver() {
+
+    val eHorizonObservers = CopyOnWriteArraySet<EHorizonObserver>()
+
+    override fun onElectronicHorizonUpdated(
+        horizon: ElectronicHorizon,
+        type: ElectronicHorizonResultType
+    ) {
+        jobController.scope.launch {
+            eHorizonObservers.forEach {
+                it.onElectronicHorizonUpdated(
+                    horizon.mapToEHorizon(),
+                    type.convertResultType()
```

Comment:
Can we do the mapping on the worker thread (or whichever thread the NN invokes this) and only notify on the main thread?

Reply:
Thought about it as well and was planning to implement as a follow up. We need to figure how to inject the `NavigatorDispatcher` from `MapboxNativeNavigatorImpl` - maybe it's as simple as making it `internal` and access to it directly from `MapboxNativeNavigatorImpl` singleton.